lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -0,0 +1,443 @@
+          continue;
+        }
+        for (uint32_t MBlock : MBlockCandidates) {
+          for (uint32_t innerMostMBlock : innerMostMBlockCandidates) {
+            if (MBlock % innerMostMBlock != 0 ||
+                shape[0] % innerMostMBlock != 0) {
+              continue;
+            }
+            for (uint32_t NBlock : NBlockCandidates) {
+              for (uint32_t innerMostNBlock : innerMostNBlockCandidates) {
+                if (NBlock % innerMostNBlock != 0 ||
+                    shape[1] % innerMostNBlock != 0) {
+                  continue;
+                }
+                for (uint32_t KBlock : KBlockCandidates) {
+                  for (uint32_t innerMostKBlock : innerMostKBlockCandidates) {
+                    if (KBlock % innerMostKBlock != 0 ||
+                        shape[2] % innerMostKBlock != 0) {
+                      continue;
+                    }
+                    MatmulConfig config{
+                        MThreads,        NThreads,        KThreads,
+                        MBlock,          NBlock,          KBlock,
+                        innerMostMBlock, innerMostNBlock, innerMostKBlock};
+                    configs.push_back(config);
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  LLVM_DEBUG(
+      llvm::dbgs() << "Finish generating candidates. ConfigCandidates size: "
+                   << configs.size() << "\n");
+  return configs;
+}
+
+// read the config from the attributes for tuning
+bool readConfigFromAttrs(MatmulConfig &config, ArrayRef<NamedAttribute> attrs) {
+  size_t cfgItemCnt = 0;
+  for (auto &attr : attrs) {
+    if (attr.getName() == "KBlock") {
+      config.KBlock = cast<IntegerAttr>(attr.getValue()).getInt();
+      cfgItemCnt++;
+    } else if (attr.getName() == "KThreads") {
+      config.KThreads = cast<IntegerAttr>(attr.getValue()).getInt();
+      cfgItemCnt++;
+    } else if (attr.getName() == "NBlock") {
+      config.NBlock = cast<IntegerAttr>(attr.getValue()).getInt();
+      cfgItemCnt++;
+    } else if (attr.getName() == "NThreads") {
+      config.NThreads = cast<IntegerAttr>(attr.getValue()).getInt();
+      cfgItemCnt++;
+    } else if (attr.getName() == "MBlock") {
+      config.MBlock = cast<IntegerAttr>(attr.getValue()).getInt();
+      cfgItemCnt++;
+    } else if (attr.getName() == "MThreads") {
+      config.MThreads = cast<IntegerAttr>(attr.getValue()).getInt();
+      cfgItemCnt++;
+    } else if (attr.getName() == "innermostMBlock") {
+      config.innerMostMBlock = cast<IntegerAttr>(attr.getValue()).getInt();
+      cfgItemCnt++;
+    } else if (attr.getName() == "innermostNBlock") {
+      config.innerMostNBlock = cast<IntegerAttr>(attr.getValue()).getInt();
+      cfgItemCnt++;
+    } else if (attr.getName() == "innermostKBlock") {
+      config.innerMostKBlock = cast<IntegerAttr>(attr.getValue()).getInt();
+      cfgItemCnt++;
+    }
+  }
+  return cfgItemCnt == 9;
```

Comment:
I was referring to validate all the fields of `config`, in case user provides any invalid number to any field of it by mistake.

Reply:
Thanks for the reminder! The check for every field and the check for index access are added.